### PR TITLE
Basic warning support implementation for the Godot Shading Language.

### PR DESCRIPTION
**Pull request by Chaosus**
Basic warning support implementation for the Godot Shading Language.

warnings

Currently implemented warnings:

- UNUSED_CONSTANT
- UNUSED_VARYING
- UNUSED_UNIFORM
- UNUSED_FUNCTION
- UNUSED_STRUCT

They are configured inside project settings:

image

### CODE REVIEW

**Comment by Calinou** [comment]
Is it possible to add a warning for floating-point comparisons (`if (i == 0.3)`)? This is probably one of the most useful warnings for shader beginners :slightly_smiling_face:

**Comment by Chaosus** [comment]
> Is it possible to add a warning for floating-point comparisons (if (i == 0.3))? This is probably one of the most useful warnings for shader beginners

I guess ), this is just a beginning I will make them gradually.

**Comment by Chaosus** [comment]
@Calinou Thx for the pointing. Added FLOAT_COMPARSION warning:

image

**Comment by Calinou** [comment]
@Chaosus Looks good, but I'd reword the message for better grammar:

`Direct flaoting-point comparison (this may not evaluate to `true` as you expect). Instead, use `abs(a -b) < 0.0001` for an approximate but predictable comparison.`

I'd also change the suggested epsilon to `0.0001` as it's more likely to work in all cases.

**Comment by Chaosus** [comment]
@Calinou Thanks, fixed.

File: servers/rendering/shader_warnings.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+const StringName &ShaderWarning::get_subject() const {
+	return subject;
+}
+
+String ShaderWarning::get_message() const {
+	switch (code) {
+		case FLOAT_COMPARISON:
+			return vformat(TTR("Direct floating-point comparison (this may not evaluate to `true` as you expect). Instead, use `abs(a -b) < 0.0001` for an approximate but predictable comparison."));
```
Comment (Calinou) [inline servers/rendering/shader_warnings.cpp:51]:
```suggestion
			return vformat(TTR("Direct floating-point comparison (this may not evaluate to `true` as you expect). Instead, use `abs(a - b) < 0.0001` for an approximate but predictable comparison."));
```

Comment (Chaosus) [reply servers/rendering/shader_warnings.cpp]: Fixed

**Comment by Chaosus** [comment]
I rebased it - It's working correctly. If you can please check and approve it @Calinou. This is basic implementation - in the future, I want to add a `UNUSED_LOCAL_VARIABLE` warning to this system, but for now, this just works and I want it to be merged. Thanks.

File: servers/rendering/shader_warnings.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+const StringName &ShaderWarning::get_subject() const {
+	return subject;
+}
+
+String ShaderWarning::get_message() const {
+	switch (code) {
+		case FLOAT_COMPARISON:
+			return vformat(TTR("Direct floating-point comparison (this may not evaluate to `true` as you expect). Instead, use `abs(a - b) < 0.0001` for an approximate but predictable comparison."));
```
Comment (Calinou) [inline servers/rendering/shader_warnings.cpp:51]: How will this behave in non-editor builds (e.g. in the console log of export templates)? `TTR()` is not available there.

**Review by Calinou — COMMENTED** [review]
Tested locally, it works:

image

In the future, we'll probably want to add a system to use "ignore comments" to ignore specific warnings in shader files. This will make the "treat warnings as errors" project setting more usable.

File: servers/rendering/shader_warnings.cpp (hunk 2 above)
Comment (akien-mga) [reply servers/rendering/shader_warnings.cpp:51]: Does the warning logic make sense outside `TOOLS_ENABLED`?
If so, this could likely use `RTR()` which works for both tools and runtime (the translation will not be available unless the users provide it themselves in their game though).

Comment (Chaosus) [reply servers/rendering/shader_warnings.cpp:51]: I also don't know, I could remove the TTR completely for now, sure.

Comment (Chaosus) [reply servers/rendering/shader_warnings.cpp]: Okay, I removed these TTR (sorry for the long delay - was busy with another PR).

File: editor/plugins/shader_editor_plugin.cpp [hunk 3]
```diff
@@ -37,12 +37,18 @@
...
 #include "editor/property_editor.h"
 #include "servers/display_server.h"
 #include "servers/rendering/shader_types.h"
 
 /*** SHADER SCRIPT EDITOR ****/
 
+static bool saved_warnings_enabled = false;
+static bool saved_treat_warning_as_errors = false;
+static Map<ShaderWarning::Code, bool> saved_warnings;
+static uint32_t saved_warning_flags = 0U;
```
Comment (akien-mga) [inline editor/plugins/shader_editor_plugin.cpp:50]: Is there a specific reason to make those `static` here instead of private members of `ShaderTextEditor`?

File: servers/rendering/shader_language.h [hunk 4]
```diff
@@ -848,9 +882,13 @@ class ShaderLanguage {
 		IDENTIFIER_LOCAL_VAR,
 		IDENTIFIER_BUILTIN_VAR,
 		IDENTIFIER_CONSTANT,
-	};
+		IDENTIFIER_MAX,
+	} last_type = IDENTIFIER_MAX;
```
Comment (akien-mga) [inline servers/rendering/shader_language.h:886]: It's equivalent but I think it looks clearer if `last_type` is defined separately, since it's a named enum:
```
IdentifierType last_type = IDENTIFIER_MAX;
```

**Review by akien-mga — COMMENTED** [review]
Needs a rebase, otherwise seems good.

File: editor/plugins/shader_editor_plugin.cpp (hunk 3 above)
Comment (Chaosus) [reply editor/plugins/shader_editor_plugin.cpp:50]: No, they will not work if I put them to class.

File: servers/rendering/shader_language.h (hunk 4 above)
Comment (Chaosus) [reply servers/rendering/shader_language.h]: Fixed

File: editor/plugins/shader_editor_plugin.cpp (hunk 3 above)
Comment (akien-mga) [reply editor/plugins/shader_editor_plugin.cpp:50]: Ah yeah you use them in both `ShaderEditor` and `ShaderTextEditor` I see.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by aaronfranke** [comment]
@Chaosus Instead of suggesting the expression `abs(a - b) < 0.0001`, would it be feasible to port `is_equal_approx` and maybe `is_zero_approx` to the shader language? We should ensure that the epsilon is appropriate for whatever float precision is being used by the shader.
